map.go

```diff
@@ -517,32 +527,41 @@ func (mi *MapIterator) Next(keyOut, valueOut interface{}) bool {
 			return false
 		}
 
-		// The user can get access to nextBytes since unmarshalBytes
-		// does not copy when unmarshaling into a []byte.
-		// Make a copy to prevent accidental corruption of
-		// iterator state.
-		copy(mi.prevBytes, nextBytes)
-		mi.prevKey = mi.prevBytes
-
 		mi.err = mi.target.Lookup(nextBytes, valueOut)
 		if IsNotExist(mi.err) {
-			// Even though the key should be valid, we couldn't look up
-			// its value. If we're iterating a hash map this is probably
-			// because a concurrent delete removed the value before we
-			// could get it. If we're iterating one of the fd maps like
-			// ProgramArray it means that a given slot doesn't have
-			// a valid fd associated.
-			// In either case there isn't much we can do, so just
-			// continue to the next key.
+			if mi.target.abi.Type == Hash {
+				// XXX: We have to audit all map types for their behaviour
+
+				// Don't use nextBytes in NextKeyBytes, since it will
+				// probably reset iteration. Instead, re-try the next key
+				// lookup. If we're lucky mi.prevKey still exists, if we're
+				// unlucky we re-start from the top. There is no way to tell.
+				continue
```

Comment:
Is this trying to protect against the following?

```
next = GetNextKey(prev)
                                       delete(next)
lookup(prev)
```

This is definitely a case worth handling.

The next case I'm interested in is what if `prev` disappears during the arbitrary library user code execution. While for this `next` case, we're mostly hedging against fairly tight race conditions or the process being swapped out in the middle of `GetNextKey` vs `Lookup`, there's a relatively small amount of code involved here - a couple of syscalls and maybe some marshaling. However once we return from this function to the user library, it's entirely up to them.

There's two cases I'm concerned about here:
1. Long delay because there's actually a lot of processing that the user wants to do; then a BPF deletion occurs, or
1. The caller actually assumes that it can delete the entry that is returned through the parameters. 

Case 1 is closely tied to the ongoing discussions so I won't go into any more depth here.

Case 2 says that if this behaviour remains in the long-term implementation, we should at least state this expectation in the API.

Reply:
> Is this trying to protect against the following?

Hmm, not sure. The code does `next = GetNextKey(prev)` and then `lookup(next)`. Since `next` becomes `prev` in the next iteration, we know that iteration will reset iff we get ENOENT and the type of the map is hash. Is that what you mean?

> There's two variations I'm concerned about here:

The problem for (1) is that we can't even say that there is a reliable way to dump. Except retrying until you don't get errAborted.

Ugh, I've not thought about (2), that's even worse than the race conditions. The hash map semantics really are garbage. Yeah, we need to document that.